### PR TITLE
Add backup status checking in schedule controller.

### LINKED ISSUE #4738: Possible race condition when removing a backup in New phase and backups overlapping
**What steps did you take and what happened:**
We noticed we have a lot of backups in `New` phase on some highly loaded clusters. We believe it might be because we have a schedule configured to take a backup every 15m(`*/15 * * * *`), but the backup itself is taking more than 15m to complete (based on the diff between `Started` and `Completed` timestamps from `velero backup describe`) and this creates an overlapping between backups.

```
❯ velero backup get | grep New | wc -l
     504

❯ velero backup get | grep New
......................
hourly-k8s-backup-20220307151512   New          0        0          <nil>                           5d ago                       <none>
hourly-k8s-backup-20220307150012   New          0        0          <nil>                           5d ago                       <none>
hourly-k8s-backup-20220307140012   New          0        0          <nil>                           5d ago                       <none>
hourly-k8s-backup-20220307124512   New          0        0          <nil>                           5d ago                       <none>
hourly-k8s-backup-20220307113012   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307110012   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307090011   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307084511   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307081511   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307080011   New          0        0          <nil>                           6d ago                       <none>
hourly-k8s-backup-20220307064511   New          0        0          <nil>                           6d ago                       <none>
.....................
```

We tried to cleanup the backups that were in the `New` phase using 
```
velero backup delete $(velero backup get | grep New | awk '{print $1}') --confirm
```
but no cleanup was actually done even if no errors were encountered during the above command, e.g.
```
..............
Request to delete backup "hourly-k8s-backup-20220308221528" submitted successfully.
The backup will be fully deleted after all associated data (disk snapshots, backup files, restores) are removed.
..............
``` 

Most probably this is because the backup-deletion plugin already removed the storage location since the schedule backup TTL is configured to 24h
```
❯ velero schedule get
NAME                STATUS    CREATED                         SCHEDULE       BACKUP TTL   LAST BACKUP   SELECTOR
hourly-k8s-backup   Enabled   2021-02-11 10:26:36 +0200 EET   */15 * * * *   24h0m0s      7m ago        <none>
```

and `velero backup describe hourly-k8s-backup-20220307064511` shows:
```
Deletion Attempts (1 failed):
  2022-03-14 13:21:13 +0200 EET: Processed
  Errors:
    backup storage location  not found
```

**What did you expect to happen:**

- Small or no backups in `New` phase.
- No backups overlapping. If a backup didn't finish until the next schedule maybe skip the next schedule
- Be able to remove backups in `New` phase even if the storage location was cleaned up by backup-deletion plugin. Maybe a `--force` flag might help.

**The following information will help us better understand what's going on**:

_If you are using velero v1.7.0+:_  
Please use `velero debug  --backup <backupname> --restore <restorename>` to generate the support bundle, and attach to this issue, more options please refer to `velero debug --help` 

_If you are using earlier versions:_  
Please provide the output of the following commands (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
- `kubectl logs deployment/velero -n velero`
- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
- `velero backup logs <backupname>`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`
```
❯ velero backup describe hourly-k8s-backup-20220307064511
Name:         hourly-k8s-backup-20220307064511
Namespace:    velero
Labels:       velero.io/schedule-name=hourly-k8s-backup

Phase:  New

Errors:    0
Warnings:  0

Namespaces:
  Included:  *
  Excluded:  default, kube-public, kube-system, monitoring, kube-node-lease, ...

Resources:
  Included:        *
  Excluded:        events, events.events.k8s.io
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:

Velero-Native Snapshot PVs:  false

TTL:  24h0m0s

Hooks:  <none>

Backup Format Version:

Started:    <n/a>
Completed:  <n/a>

Expiration:  <nil>

Velero-Native Snapshots: <none included>

Deletion Attempts (1 failed):
  2022-03-14 13:21:13 +0200 EET: Processed
  Errors:
    backup storage location  not found
```

**Anything else you would like to add:**
Using azure plugin

**Environment:**

- Velero version (use `velero version`): 
```
❯ velero version
Client:
	Version: v1.8.0
	Git commit: -
Server:
	Version: v1.7.1
```

- Velero features (use `velero client config get features`): 
```
❯ velero client config get features
features: <NOT SET>
```

- Kubernetes version (use `kubectl version`):
```
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:30:48Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.11", GitCommit:"27522a29febbcc4badac257763044d0d90c11abd", GitTreeState:"clean", BuildDate:"2021-09-15T19:16:25Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version: `v1.20`
- Cloud provider or hardware configuration: Azure
- OS (e.g. from `/etc/os-release`): Flatcar

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by blackpiglet**
Signed-off-by: Xun Jiang <blackpiglet@gmail.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #4738 

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/controller/schedule_controller.go [hunk 1]
```diff
@@ -176,35 +181,61 @@ func parseCronSchedule(itm *velerov1.Schedule, logger logrus.FieldLogger) (cron.
...
+	isDue, nextRunTime := getNextRunTime(schedule, cronSchedule, c.clock.Now())
+	log := c.logger.WithField("schedule", kubeutil.NamespaceAndName(schedule))
 
 	if !isDue {
 		log.WithField("nextRunTime", nextRunTime).Debug("Schedule is not due, skipping")
-		return nil
+		return false
 	}
 
+	log.WithField("nextRunTime", nextRunTime).Info("Schedule is due, going to submit backup.")
```
Comment (ywk253100) [inline pkg/controller/schedule_controller.go:220]: Please remove the `going to submit backup` part or move it into the `submitBackup` function as there is another checking before submitting the backup

File: pkg/controller/schedule_controller.go [hunk 2]
```diff
@@ -127,10 +128,14 @@ func (c *scheduleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+	// Check for the schedule being due to run.
+	// If there are backup created by this schedule still in New or InProgress state,
+	// skip current backup creation to avoid running overlap backups.
 	// As the schedule must be validated before checking whether it's due, we cannot put the checking log in Predicate
-	if err := c.submitBackupIfDue(ctx, schedule, cronSchedule); err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "error running submitBackupIfDue for schedule %s", req.String())
+	if c.ifDue(schedule, cronSchedule) && !c.checkIfBackupInNewOrProgress(schedule) {
+		if err := c.submitBackup(ctx, schedule); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "error submit backup for schedule %s", req.String())
+		}
```
Comment (ywk253100) [inline pkg/controller/schedule_controller.go:138]: How about adding a log to explain why the schedule doesn't trigger a new backup?

File: pkg/controller/schedule_controller.go (hunk 1 above)
Comment (blackpiglet) [reply pkg/controller/schedule_controller.go:220]: Removed.

File: pkg/controller/schedule_controller.go (hunk 2 above)
Comment (blackpiglet) [reply pkg/controller/schedule_controller.go:138]: Added in function checkIfBackupInNewOrProgress

**Review by ywk253100 — APPROVED** [review]

**Review by qiuming-best — APPROVED** [review]

**Pull request merged**